GSS Auth Implementation

AndrewJackson2020 opened the pull request:
This PR implements GSS authentication in pgbouncer. This is a very important institutional feature that allows a client to verify its identity to a server (and vice versa) without the client ever sending its password to the server.

Since the underlying gss c functions are network blocking and synchronous I used the threaded design common to the PAM implementation and the LDAP PR #731 use for this functionality. For the CI and testing aspects of this PR I found the GSS backend encryption PR #743 helpful as well.

Sources:
- [GSS auth postgres docs](https://www.postgresql.org/docs/current/gssapi-auth.html)
- [Libpq backend GSS implementation](https://github.com/postgres/postgres/blob/master/src/backend/libpq/auth.c)

JelteF commented on doc/config.md:
```diff
@@ -527,6 +527,8 @@ Database name in the `[database]` section to be used for authentication purposes
 option can be either global or overridden in the connection string if this parameter is
 specified.
 
+### krb_server_keyfile
```
I guess you forgot to fill in the docs

AndrewJackson2020 replied:
sorry about that will fix.

AndrewJackson2020 replied:
Added documentation to the PR.

JelteF commented:
Just as with the LDAP PR it would be very helpful if someone that uses GSS could try this PR out. I can do a review to see if the code looks sensible, but I don't really know enough about GSS to do an indepth review of this (nor do I have the time to learn).

JelteF commented:
FYI while cleaning up old PRs I found this one: https://github.com/pgbouncer/pgbouncer/pull/441

It would probably be good to check if there are any good ideas in there.

AndrewJackson2020 commented:
Expanded the test coverage and feature set. Tried to match all of the GSS options that are available in postgres (case insensitive match, realm match, etc). I believe that I have covered currently with the exception of user mapping.

zhaorui20005 commented on src/client.c:
```diff
@@ -151,6 +152,10 @@ static bool send_client_authreq(PgSocket *client)
 		SEND_generic(res, client, PqMsg_AuthenticationRequest, "iss", AUTH_REQ_SASL, "SCRAM-SHA-256", "");
+#ifdef HAVE_GSS
+	} else if (auth_type == AUTH_TYPE_GSS) {
+		SEND_generic(res, client, 'R', "i", AUTH_REQ_GSS);
```
Better to use PqMsg_AuthenticationRequest for value 'R'.

AndrewJackson2020 replied:
Yeah this is a good point.

zhaorui20005 reviewed (commented):
I believe this is another quite important feature. Hope I can provide some helps.

AndrewJackson2020 commented on src/gss.c:
```diff
@@ -0,0 +1,594 @@
+
+	gss_first_free_slot = next_free_slot;
+
+	pthread_mutex_unlock(&gss_queue_tail_mutex);
```
This should be fixed now.

AndrewJackson2020 commented on src/gss.c:
```diff
@@ -0,0 +1,594 @@
+		}
+
+		if (gss_check_passwd(request)) {
+			request->status = GSS_STATUS_SUCCESS;
```
This should be fixed now. Went ahead and tried to implement this as closely as possible to your LDAP PR.

zhaorui20005 commented:
> I believe this is another quite important feature. Hope I can provide some helps.

I will continue reading the logic of GSS as it differs a lot for LDAP, I need sometime to compare this with postgres when I got time. And maybe test on my local vm.

AndrewJackson2020 commented on src/gss.c:
```diff
@@ -0,0 +1,619 @@
+		 * might change exactly after it), but it helps to quickly filter out invalid
+		 * sockets and thus save some time.
+		 */
+		if (!is_valid_socket(request)) {
```
Does this issue apply to your LDAP PR as well or did you do something there to mitigate the issue?

zhaorui20005 replied:
I just removed the calling here as I think this is not a must check.

AndrewJackson2020 replied:
Okay I similarly removed the above mentioned call to `is_valid_socket`. This sounds like another item that should be implemented in a separate PR for PAM as well.

AndrewJackson2020 commented:
Just a heads up: opened PR #1275 in order to keep track of some of these thread safety concerns on the PAM side.

FF-Ninja commented:
Hi, can I ask what the current state of this feature is? I am interested in using this

AndrewJackson2020 commented:
> Hi, can I ask what the current state of this feature is? I am interested in using this

The  state of this is "it works on my machine". Right now this PR needs reviews, feedback, and testing from people who are interested in this feature. Please feel free to build this from source, test it for your use case and report any issues that you find, would be greatly appreciated.

FF-Ninja commented:
> > Hi, can I ask what the current state of this feature is? I am interested in using this
> 
> The state of this is "it works on my machine". Right now this PR needs reviews, feedback, and testing from people who are interested in this feature. Please feel free to build this from source, test it for your use case and report any issues that you find, would be greatly appreciated.

Great thanks. Will build and test it

AndrewJackson2020 commented:
> Great thanks. Will build and test it

Appreciate it, thank you.

zhaorui20005 commented on test/conftest.py:
```diff
@@ -17,6 +20,75 @@
+MASTER_PASSWORD = "master_password"
+
+
+def setup_krb():
```
Where is this function called?

AndrewJackson2020 replied:
Its probably not called anymore. I originally set up kerberos in python during tests but decided python tests shouldn't do anything with su, that should all be on the container setup. Will remove.

AndrewJackson2020 replied:
Went ahead and removed the unused test assets from this file.

AndrewJackson2020 commented on src/objects.c:
```diff
@@ -45,6 +45,14 @@ struct AATree user_tree;
+ * and because gss_user_tree is closer to PgDatabase.user_tree in
+ * logic.
+ */
+struct AATree gss_user_tree;
```
Huh, not sure why I did it like that. Will review what you did in LDAP and what exists with PAM and will modify this accordingly.

AndrewJackson2020 commented on src/gss.c:
```diff
@@ -0,0 +1,608 @@
+	delegated_creds = GSS_C_NO_CREDENTIAL;
+
+	/* Accept the context. */
+	maj_stat = gss_accept_sec_context(&acc_sec_min_stat,
```
Correct, good catch.

AndrewJackson2020 replied:
This should be fixed after my refactor.

AndrewJackson2020 commented:
@zhaorui20005 just a heads up, I'm going to try to rewrite the GSS part of this PR entirely to make it as similar as possible to Postgres's GSS implementation. I think this will address 3 of your comments above and make the implementation more maintainable.

AndrewJackson2020 commented on src/gss.c:
```diff
@@ -0,0 +1,608 @@
+		  maj_stat, min_stat,
+		  (unsigned int) send_tok.length, *request->flags);
+
+	if (GSS_S_COMPLETE != maj_stat) {
```
This should be good after my refactor.

AndrewJackson2020 commented on src/gss.c:
```diff
@@ -0,0 +1,608 @@
+				}
+			}
+		}
+	}
```
I restructured these if statements to look a lot more like how it functions in the postgres source code. I think this issue should be fixed now.

AndrewJackson2020 commented:
@zhaorui20005 I have completed my response to all of your feedback. Tests are all passing. The entire interaction with the GSS API should now look a lot more like the postgres source code which I believe will make it much easier to review. 

When you get a chance please take a look and let me know if you have additional feedback.

zhaorui20005 commented on src/gss.c:
```diff
@@ -0,0 +1,607 @@
+	server_credentials = GSS_C_NO_CREDENTIAL;
+	delegated_creds = GSS_C_NO_CREDENTIAL;
+
+	/* Accept the context. */
```
The overall logic looks fine for me. 
But only one question from here, in Postgres, there is a loop keep calling function `gss_accept_sec_context ()`. But here there is only one round, can you explain why one round of calling is enough, and in which case there will be multiple round of exchange between the client and server?

AndrewJackson2020 replied:
I cannot explain in what case multiple rounds would be required. All of my testing works with only a single round. I imagine it could have to do with a server configuration. I was a hesitant adding the loop without ever having seen a test case where the second iteration would be triggered but I could just go ahead and add the loop anyway to confirm with the postgres implementation.

zhaorui20005 replied:
I think there is no need to add the loop if we know the loop is not really needed. Better to know this in more details before make the changes. 
The first time I read the code of Postgres's implementation, I was confused about the loop too.

AndrewJackson2020 replied:
So looking at my code it looks like it assumes the auth is complete when `maj_stat == GSS_S_CONTINUE_NEEDED` whereas in the postgres codebase it goes on to do another iteration until `maj_stat == GSS_S_COMPLETE`. In manual testing I do not see instances where it actually does want to continue. I think I need to understand what causes gss to want to continue in order to move forward with this PR.

zhaorui20005 replied:
So it is better to add that loop for better portablility?

AndrewJackson2020 replied:
Sounds like it, though I'm not sure exactly what portability means in this context. Also I think if we were to modify this PR to enable the multiple rounds of token exchange it wouldn't be a loop locally in the GSS.c file but it would return to the main sequence. This makes me more concerned about implementing it without test coverage.

zhaorui20005 replied:
So you are worried about the loop may block the whole working thread?
If you did not add that loop, maybe the authentication should fail when return value is `GSS_S_CONTINUE_NEEDED `.
But I think it is better to add that loop, as we now do not know in which case the loop may happen or even block.

zhaorui20005 replied:
I re-read the code, it is difficult for pgbouncer to receive the token in gss.c
So I am OK with current state of the code now.
If there are bad cases reported, then we can think of other ways to add that loop.

AndrewJackson2020 replied:
Okay sounds good. I do think that a return of `GSS_S_CONTINUE_NEEDED` should lead to a failure though. Will update this next week.

zhaorui20005 replied:
In the current function of `gss_recvauth()`. There are 3 resources that needs to be cleared after return, as without the loop, some resources will be left uncleared.
The `server_credentials`,calling of `gss_delete_sec_context` and `gss_release_buffer(&lmin_s, &send_tok)`.
The release of `send_tok` is handled properly, but "`server_credentials`,calling of `gss_delete_sec_context`" are not, as there are return statement before clearing these 2 resources.
So I suggest to handle them properly in your next update.

AndrewJackson2020 replied:
So I modified the function so that if `GSS_S_CONTINUE_NEEDED` is received then authentication is failed. I think it is better to do it this way, at least until we figure out how to incorporate the continue procedure and get a test to trigger it.

Regarding the memory deallocation, I tried to follow the postgres code as closely as possibly with this. Is this an issue with postgres as well? Also maybe we should consider adding gss (and maybe ldap) to the valgrind test pipeline?

zhaorui20005 replied:
It is better to have a valgrind testing result if possible.

zhaorui20005 commented:
> @zhaorui20005 I have completed my response to all of your feedback. Tests are all passing. The entire interaction with the GSS API should now look a lot more like the postgres source code which I believe will make it much easier to review.
> 
> When you get a chance please take a look and let me know if you have additional feedback.

The logic looks fine to me now. I think the next step should request @JelteF to review on the basis of these changes.

JelteF commented:
FYI this has a bunch of merge conflicts now

JelteF commented:
Also same change as in #1297 should be applied here

AndrewJackson2020 commented:
> Also same change as in #1297 should be applied here
Changed the log_debug to log_warning.

> FYI this has a bunch of merge conflicts now
These were all due to the LDAP PR getting merged. These merge conflicts are all resolved now.

zhaorui20005 commented on src/gss.c:
```diff
@@ -0,0 +1,607 @@
+	request->status = GSS_STATUS_IN_PROGRESS;	/* This is protected by gss_queue_tail_mutex */
+	memcpy(&request->remote_addr, &client->remote_addr, sizeof(client->remote_addr));
+	safe_strcpy(request->username, client->login_user_credentials->name, MAX_USERNAME);
+// safe_strcpy(request->password, passwd, MAX_PASSWORD);
```
The password is not needed, so we should remove this line?

AndrewJackson2020 replied:
Correct, just pushed an update with this line removed.

zhaorui20005 commented on src/gss.c:
```diff
@@ -0,0 +1,606 @@
+		return false;
+	}
+
+	princ = malloc(gbuf.length + 1);
```
After a little testing, I found that the memory that `princ` points to be set to 0, otherwise it will fail because of the last character not match.

AndrewJackson2020 replied:
Okay, that's not good. I thought that I tested every case around this. Will take another look.

AndrewJackson2020 replied:
I changed this from malloc (uninitialized) to calloc (zero initialized). Let me know if you think this works.

zhaorui20005 replied:
Yes, it works.

zhaorui20005 commented on src/client.c:
```diff
@@ -588,6 +624,20 @@ bool set_pool(PgSocket *client, const char *dbname, const char *username, const
+		if (!check_db_connection_count(client))
+			return false;
+		if (!client->login_user_credentials) {
+			slog_error(client, "set_pool(): failed to allocate new PAM user");
```
`PAM` to `GSS`

AndrewJackson2020 replied:
This is fixed now.

zhaorui20005 commented on src/client.c:
```diff
@@ -1359,6 +1409,25 @@ static bool handle_client_startup(PgSocket *client, PktHdr *pkt)
+			 * Get the token that the client sends as part of the context
+			 * initialization process.
+			 */
+			if (!mbuf_get_bytes(&pkt->data, length, (const uint8_t **)&token)) {
```
If no disconnect_client() is called, how to handle this error in the following procedure?

AndrewJackson2020 replied:
I'm not sure about this one to be honest. Looking at the other calls in `handle_client_startup` 2 of them do not call disconnect_client explicitly and one of them does.

zhaorui20005 commented on include/gss.h:
```diff
@@ -0,0 +1,42 @@
+
+/*
+ * Defines how many authentication requests can be placed to the waiting queue.
+ * When the queue is full calls to pam_auth_begin() will block until there is
```
`pam` to `gss`

AndrewJackson2020 replied:
Good catch will fix this.

AndrewJackson2020 replied:
This is fixed now.

zhaorui20005 commented on src/gss.c:
```diff
@@ -0,0 +1,607 @@
+		if (!pktbuf_send_immediate(&_buf, request->client)) {
+			gss_release_buffer(&lmin_s, &send_tok);
+			log_debug("Failed to send gss token to client");
+			return false;
```
In postgres, there is no such return. So if this return happens, I think there is a need to call `gss_delete_sec_context(&lmin_s, &request->context, GSS_C_NO_BUFFER);` and `gss_release_cred(&min_stat, &server_credentials);` if the if block is met.

zhaorui20005 commented on src/gss.c:
```diff
@@ -0,0 +1,607 @@
+		gss_delete_sec_context(&lmin_s, &request->context, GSS_C_NO_BUFFER);
+		log_debug("accepting GSS security context failed major: %u, minor: %u",
+			  maj_stat, min_stat);
+		return false;
```
The same necessary checking for calling `gss_release_cred(&min_stat, &server_credentials);` if the return happens. Or if this is not met, then there is no need to call `gss_release_cred(&min_stat, &server_credentials);` ?
I am not 100% sure for that.

AndrewJackson2020 commented:
> @JelteF @zhaorui20005 Is there any way I can help to add this feature in the pgbouncer? This would be great to add. Thanks!

Any help testing would be greatly appreciated. Looks like this PR needs a rebase as well but thats on me.

zhaorui20005 commented:
> > @JelteF @zhaorui20005 Is there any way I can help to add this feature in the pgbouncer? This would be great to add. Thanks!
> 
> Any help testing would be greatly appreciated. Looks like this PR needs a rebase as well but thats on me.

I would like to give an approval to this PR. I have tested with the testing cases, it all passed.

AndrewJackson2020 commented:
> > @JelteF @zhaorui20005 Is there any way I can help to add this feature in the pgbouncer? This would be great to add. Thanks!
> 
> Any help testing would be greatly appreciated. Looks like this PR needs a rebase as well but thats on me.

This has been rebased.

AndrewJackson2020 commented:
> > > @JelteF @zhaorui20005 Is there any way I can help to add this feature in the pgbouncer? This would be great to add. Thanks!
> > 
> > 
> > Any help testing would be greatly appreciated. Looks like this PR needs a rebase as well but thats on me.
> 
> I would like to give an approval to this PR. I have tested with the testing cases, it all passed.

The major reservation that I still have with this architecture is that the man page for GSS says that the auth should be run in a loop with the expectation of multiple back and forth messages. In testing I have never been able trigger the necessity of the extra message. I also am not sure how this loop type flow could be implemented in the pgbouncer async framework, all of the other auth types are much more open/close compared to GSS.

[0]: https://man.freebsd.org/cgi/man.cgi?query=gss_accept_sec_context&sektion=3&apropos=0&manpath=FreeBSD+7.1-RELEASE

zhaorui20005 commented:
> > > > @JelteF @zhaorui20005 Is there any way I can help to add this feature in the pgbouncer? This would be great to add. Thanks!
> > > 
> > > 
> > > Any help testing would be greatly appreciated. Looks like this PR needs a rebase as well but thats on me.
> > 
> > 
> > I would like to give an approval to this PR. I have tested with the testing cases, it all passed.
> 
> The major reservation that I still have with this architecture is that the man page for GSS says that the auth should be run in a loop with the expectation of multiple back and forth messages. In testing I have never been able trigger the necessity of the extra message. I also am not sure how this loop type flow could be implemented in the pgbouncer async framework, all of the other auth types are much more open/close compared to GSS.

The whole logic looks fine for me. I am ok with this PR.

karankhanchandani commented:
@JelteF @zhaorui20005 Thanks for the approvals. Is it possible to merge this feature? Please let me know if I can be of any help.. Thanks!

AndrewJackson2020 commented:
> @JelteF @zhaorui20005 Thanks for the approvals. Is it possible to merge this feature? Please let me know if I can be of any help.. Thanks!

There is a memory error in this PR that I still need to figure out.

AndrewJackson2020 commented:
> > @JelteF @zhaorui20005 Thanks for the approvals. Is it possible to merge this feature? Please let me know if I can be of any help.. Thanks!
> 
> There is a memory error in this PR that I still need to figure out.

The memory error in question appears to also be an issue with LDAP and PAM. The only reason CI passes is because those code paths aren't tested with Valgrind. Went ahead and removed the Valgrind check from this PR, will open an issue for existing auth types later with more detail in case this needs to be addressed for existing auth types.

Commit:
Implement GSS authentication

This commit introduces GSS authentication to pgbouncer. Specifically
this implements authentication from clients to pgbouncer, pgbouncer to
postgres GSS authentication is out of scope of this commit. This
implementation is largely based on the implementation in postgres [0],
prior attempts at merging GSS functionality into pgbouncer [1][2], and
pgbouncer LDAP/PAM functionality. Also please note that this commit does
not include GSSAPI Encryption[3], only GSSAPI authentication, encryption
can be included as an additional future feature if requested. This
commit includes both non-HBA and HBA usages of GSSAPI, similar to LDAP
and PAM implementations in pgbouncer.

The majority of the implmentation details are in src/gss.c. The
structure of the implmentation of this file is a mirror of `src/pam.c`
and `src/ldapauth.c` but with the GSS implementation in `src/interfaces/libpq/fe-auth.c`
in the postgres source tree. Please note that one difference between the
implementation in this commit and the implemenation in postgres and what
is recommended in the FreeBSD man pages[4]: "Thus a typical portable caller
should always invoke gss_accept_sec_context() within a loop".
Per the IETF RFC [5] mentioned in the postgres docs [6]
GSS_S_CONTINUE_NEEDED can show up in the defualt mechanism with "Kerberos
V5 Double TGT".

In testing in local and in real environments I have never seen  gss_accept_sec_context return
GSS_S_CONTINUE_NEEDED within the postgres authentication flow. Because
of this I have decided to not support this. Unfortunately with
pgbouncers thread based auth methods it would not be as simple as a loop
as seen in the BSD man pages [4] as it would need to cache the state,
get another message from the client in the main thread and return to the
auth worker, as such I am very hesitant to implement this without being
able to recreate the conditions that would facilitate GSS_S_CONTINUE_NEEDED return.

Added configuration options:
- auth_krb_server_keyfile
  - equivalent to postgres krb_server_keyfile GUC [7]
- auth_krb_caseins_users
  - analogous to postgres krb_caseins_users GUC [8]
- auth_gss_parameter
  - equivalent to postgres gss HBA options [6]

The postgres GUC parameter gss_accept_delegation [9] is not included in this
implementation. This functionality is only used in postgres for dblink/postgres_fdw
to allow the user to use the same authentication from client to postgres as
from postgres to postgres. Because we do not implement pgbouncer to postgres
authentication it does not make sense to include this commit. This may make sense
to include in a future commit with pgbouncer to postgres GSS auth but it may
also not make sense given the multiplexing nature of pooler client to server connections.

Another postgres GSS feature not implemented in this feature is user
mapping. This can be added as an additional feature in the future if
requested.

[0]: https://github.com/postgres/postgres/blob/master/src/interfaces/libpq/fe-auth.c#L50
[1]: https://github.com/pgbouncer/pgbouncer/pull/441
[2]: https://github.com/pgbouncer/pgbouncer/pull/743
[3]: https://www.postgresql.org/docs/current/gssapi-enc.html
[4]: https://man.freebsd.org/cgi/man.cgi?query=gss_accept_sec_context&sektion=3&apropos=0&manpath=FreeBSD+7.1-RELEASE
[5]: https://datatracker.ietf.org/doc/html/rfc2743#page-89
[6]: https://www.postgresql.org/docs/current/gssapi-auth.html
[7]: https://www.postgresql.org/docs/current/runtime-config-connection.html#GUC-KRB-SERVER-KEYFILE
[8]: https://www.postgresql.org/docs/current/runtime-config-connection.html#GUC-KRB-CASEINS-USERS
[9]: https://www.postgresql.org/docs/current/runtime-config-connection.html#GUC-GSS-ACCEPT-DELEGATION

AndrewJackson2020 committed:
Merge branch 'master' into gss
